### ISSUE TITLE
Not running under iOS and Ionic when using "use:'file'"

### ISSUE THREAD

**Issue by Martin-Luckow** [VirtuoWorks/CanvasCameraPlugin#10]
II ran into problems with an Ionic app and the plugin: when running under android the world is fine, but under iOS and using "use:'file'" I get the console-error:
[Error] Not allowed to load local resource: file:///var/mobile/Containers/Data/Application/CFD416B7-AD48-48A9-9511-AA10E19F92B7/Documents/tmp/f16-canvascamera.jpg?seed=1264145267490218
	(anonyme Funktion) (CanvasCamera.js:200)
	(anonyme Funktion)
	runTask (polyfills.js:3:10845)
	invokeTask (polyfills.js:3:16802)
	n (polyfills.js:3)

I copied your (working) cordova-example into the blank Ionic starter app and got the same problem.
All whitelisting tricks and a full open security policy don't help.

Using "use:'data'" works is a workaround, but melting down my poor 5S :)

**Comment by Sebastien-Guillon** [comment]
Hi Dr. Martin Luckow,

Sorry we can't address these issues immediately, we will look at it soon 
as we also has some projects depending on the plugin.

Thank You for your feedback.


Sébastien Guillon,



Le 07/01/2018 à 17:44, Dr. Martin Luckow a écrit :

**Comment by Sami-Radi** [comment]
Hello,

The latest cordova-ios application uses wkwebview instead of webview. WKWebView doesn't allow local resource files loading (i.e. URLs such as file://...) from the loaded html file. However there are workarounds. I'll update you as soon as i find a suitable solution.

Best Regards,

**Comment by Martin-Luckow** [comment]
Thanks for your answer.

Btw... I have a question and kind of a feature request :)
Some ugly phones (ynwim) show a horizonatally front cam preview, i.e. you look into a mirror. Others show unflipped previews... closer to reality.

In my app I use the front cam for real time pattern recognition and have to know if the image is flipped or not. Currently, the user has to check it by himself in the apps configuration.

Do you see a chance to detect flipped video streams on the level of the plugin?

Best regards

Martin

**Comment by Sami-Radi** [comment]
Hello,

Indeed in fact this plugin is a light version of 2 plugins we designed using : 

- openCV (C++ lib) for pattern recognition on the native side;
- and another one using the native API for facial detection;

Each of These two plugins solve the problem by processing the raw image. However we are not planning on releasing them as open source projects yet. 

Initially we used trackingjs but poor performances led us to relying on the native side for this kind of heavy processing. 

Best regards,

**Comment by Sami-Radi** [comment]
Hello, 

In the latest release the mirroring problem is fixed. However, I could not reproduce the bug you reported.
I implemented a fix, it works fine with our phone. Could you test the latest release and tell us if the problem still occurs ("Not allowed to load local resource...").

If it does, could you give us the iOS version you're using and the phone model.

Thanks, 

Best regards

Sami

**Comment by Sami-Radi** [comment]
Hello,

Any news ?

Best Regards,

**Comment by Martin-Luckow** [comment]
Hi Sami.

First: sorry for the late answer.
After some iOS-updates I tried it today with the latest version of your plugin and the option use:"file". The problem remains.
The device is an iPhone 5s with iOS 11.4.1.

Best regards

Martin

![bildschirmfoto 2018-08-30 um 09 48 33](https://user-images.githubusercontent.com/11980340/44837437-ea5ee100-ac39-11e8-99a4-61a92e4fd8b4.png)

**Comment by Martin-Luckow** [comment]
Hi Sami.

I found the reason:
When running under iOS "use":"file" works only if using UIWebView.
Using the newer WKWebView leads to the "Not allowed"-error above. Seems, that this could be solved by rewriting "file://..." to "//...". 
(My "fault": I am using Ionic, which uses WKWebView by default.)

Annother impact: using "use":"data" has the effect, that on a iPhone 5s the machine/app gets slower and slower and exits finally (after 20 minutes or so). Meanwhile the poor iPhone is close to its melting point :)
I have the feeling, the garbage collection is not fast enough to dispose outdated images.
This problem does not exist when using "file".

It woud be great, if you could detect UIWebView or WKWebView in your plugin and create the file-url accordingly.

Best Regards

Martin

**Comment by BHuys** [comment]
Is there any progress on this issue? I'm having the same permission problem with WKWebView, using data kills my performance.

**Comment by duncan-qabana** [comment]
Hey Sami,

I saw your comment on #15 . I just wanted to let you know that it also doesn't work anymore on ios 12.2.
After updating my iphone 6s to 12.2, the plugin constantly throws this error. Using data is not really an option for the app I am developing.

If you need anymore information. I am available to provide it.

**Comment by Sami-Radi** [comment]
Hello,

It seems that Ionic is using as a default the [Ionic Web View for Cordova plugin](https://github.com/ionic-team/cordova-plugin-ionic-webview). According to its [documentation](https://ionicframework.com/docs/v3/wkwebview/) the plugin changes Cordova default UIWebView to WKWebView. 

According to previous messages :

> When running under iOS "use":"file" works only if using UIWebView.
> Using the newer WKWebView leads to the "Not allowed"-error above. Seems, that this could be solved by rewriting "file://..." to "//...".
> ...
> It woud be great, if you could detect UIWebView or WKWebView in your plugin and create the file-url accordingly.

The latest version of the plugin implements exactly this solution. Checks if the cordova-plugin-ionic-webview plugin is in use, and rewrite the URL accordingly.

In conclusion, the bug should be fixed with the latest version.

Best regards.

**Comment by Sami-Radi** [comment]
Hello,

I believe the problem is fixed now. Closing the issue.

Best Regards.
